access-control-app/src/cmr/access_control/services/acl_service.clj

```diff
@@ -357,3 +360,39 @@
     (if (or (nil? user-token)(re-matches #"^[\s\t]*$" user-token))
       (auth-util/get-sids context :guest)
       (auth-util/get-sids context (tokens/get-user-id context user-token)))))
+
+(defn- fetch-s3-buckets-by-groups
+  [context groups providers]
+  (let [base-query (gc/or-conds (map #(qm/string-condition :permitted-group-ids
+                                                           (format "*%s*" %)
+                                                           :false
+                                                           :true) groups))
+        condition (if (empty? providers)
+                    base-query
+                    (gc/and
+                     base-query
+                     (gc/or-conds (map #(qm/string-condition :provider-id %) providers))))
+        query (qm/query {:concept-type :collection
+                         :condition condition
+                         :skip-acls? true
+                         :page-size :unlimited
+                         :result-format :query-specified
+                         :result-fields [:s3-bucket-and-object-prefix-names]})
+        es-results (qe/execute-query context query)]
+    (->> es-results
+         :items
+         (map :s3-bucket-and-object-prefix-names)
+         flatten
+         set
+         sort)))
+
+(defn s3-buckets-for-user
+  "Returns a list of s3 buckets and object prefix names by provider."
+  [context user provider-ids]
+  (let [groups (auth-util/get-sids context user)
+        providers (if (empty? provider-ids)
+                    (map :provider-id (mdb/get-providers context))
```

Comment:
providers should be cached in access-control. We can't afford to call metadata-db to retrieve providers on every get s3 prefix call that does not specify a provider.